Index: component/webdav/src/main/java/org/exoplatform/services/jcr/webdav/WebDavServiceImpl.java
===================================================================
--- component/webdav/src/main/java/org/exoplatform/services/jcr/webdav/WebDavServiceImpl.java	(revision 79046)
+++ component/webdav/src/main/java/org/exoplatform/services/jcr/webdav/WebDavServiceImpl.java	(working copy)
@@ -1008,6 +1008,9 @@
         if (element.contains("'")) {
           element = element.replaceAll("'", URLEncoder.encode("'", "UTF-8"));
         }
+        else if (element.contains("&")) {
+          element = element.replaceAll("&", URLEncoder.encode("&", "UTF-8"));
+        }
         escapedPath.append(element + "/");
       } catch (Exception e) {
         log.warn(e.getMessage());
