Index: component/ext/src/main/java/org/exoplatform/services/jcr/ext/replication/ProxyWorkspaceDataReceiver.java
===================================================================
--- component/ext/src/main/java/org/exoplatform/services/jcr/ext/replication/ProxyWorkspaceDataReceiver.java	(revision 63243)
+++ component/ext/src/main/java/org/exoplatform/services/jcr/ext/replication/ProxyWorkspaceDataReceiver.java	(working copy)
@@ -19,6 +19,7 @@
 import org.exoplatform.services.jcr.config.RepositoryConfigurationException;
 import org.exoplatform.services.jcr.impl.core.lock.LockManagerImpl;
 import org.exoplatform.services.jcr.impl.core.query.SearchManager;
+import org.exoplatform.services.jcr.impl.core.query.SystemSearchManager;
 import org.exoplatform.services.jcr.impl.dataflow.persistent.CacheableWorkspaceDataManager;
 
 /**
@@ -42,7 +43,7 @@
    */
   public ProxyWorkspaceDataReceiver(CacheableWorkspaceDataManager dataManager,
                                     LockManagerImpl lockManager) throws RepositoryConfigurationException {
-    this(dataManager, null, lockManager);
+      this(dataManager, null, lockManager, null);
   }
 
   /**
@@ -57,7 +58,7 @@
    */
   public ProxyWorkspaceDataReceiver(CacheableWorkspaceDataManager dataManager,
                                     SearchManager searchManager) throws RepositoryConfigurationException {
-    this(dataManager, searchManager, null);
+      this(dataManager, searchManager, null, null);
   }
 
   /**
@@ -69,7 +70,7 @@
    *           will be generated RepositoryConfigurationException
    */
   public ProxyWorkspaceDataReceiver(CacheableWorkspaceDataManager dataManager) throws RepositoryConfigurationException {
-    this(dataManager, null, null);
+      this(dataManager, null, null, null);
   }
 
   /**
@@ -86,7 +87,9 @@
    */
   public ProxyWorkspaceDataReceiver(CacheableWorkspaceDataManager dataManager,
                                     SearchManager searchManager,
-                                    LockManagerImpl lockManager) throws RepositoryConfigurationException {
-    dataKeeper = new WorkspaceDataManagerProxy(dataManager, searchManager, lockManager);
+            LockManagerImpl lockManager, SystemSearchManager systemSearchIndexer)
+            throws RepositoryConfigurationException
+   {
+      dataKeeper = new WorkspaceDataManagerProxy(dataManager, searchManager, lockManager, systemSearchIndexer);
   }
 }
Index: component/ext/src/main/java/org/exoplatform/services/jcr/ext/replication/WorkspaceDataManagerProxy.java
===================================================================
--- component/ext/src/main/java/org/exoplatform/services/jcr/ext/replication/WorkspaceDataManagerProxy.java	(revision 63243)
+++ component/ext/src/main/java/org/exoplatform/services/jcr/ext/replication/WorkspaceDataManagerProxy.java	(working copy)
@@ -28,6 +28,7 @@
 import org.exoplatform.services.jcr.dataflow.persistent.ItemsPersistenceListener;
 import org.exoplatform.services.jcr.impl.core.lock.LockManagerImpl;
 import org.exoplatform.services.jcr.impl.core.query.SearchManager;
+import org.exoplatform.services.jcr.impl.core.query.SystemSearchManager;
 import org.exoplatform.services.jcr.impl.dataflow.persistent.CacheableWorkspaceDataManager;
 import org.exoplatform.services.log.ExoLogger;
 
@@ -65,13 +66,17 @@
    */
   public WorkspaceDataManagerProxy(CacheableWorkspaceDataManager dataManager,
                                    SearchManager searchIndex,
-                                   LockManagerImpl lockManager) {
+                                   LockManagerImpl lockManager,
+                                   SystemSearchManager systemSearchIndexer) {
     this.listeners = new ArrayList<ItemsPersistenceListener>();
     listeners.add(dataManager.getCache());
     if (searchIndex != null)
       listeners.add(searchIndex);
     if (lockManager != null)
       listeners.add(lockManager);
+    if (systemSearchIndexer != null)
+      listeners.add(systemSearchIndexer);
+    
     log.info("WorkspaceDataManagerProxy is instantiated");
   }
 
@@ -87,7 +92,7 @@
   public void save(ItemStateChangesLog changesLog) throws InvalidItemStateException,
                                                   UnsupportedOperationException,
                                                   RepositoryException {
-    for (ItemsPersistenceListener listener : listeners) {
+    for (ItemsPersistenceListener listener : listeners) { 
       listener.onSaveItems(changesLog);
     }
     if (log.isDebugEnabled())
