Index: component/rmi/src/main/java/org/exoplatform/services/jcr/rmi/api/value/PermissionValue.java
===================================================================
--- component/rmi/src/main/java/org/exoplatform/services/jcr/rmi/api/value/PermissionValue.java	(revision 50885)
+++ component/rmi/src/main/java/org/exoplatform/services/jcr/rmi/api/value/PermissionValue.java	(working copy)
@@ -36,6 +36,8 @@
 public class PermissionValue extends BaseNonStreamValue implements Serializable, StatefulValue {
 
   private static final long serialVersionUID = 1475495561074957852L;
+  
+  private static final String IDENTITY_DELIMITER = " ";
 
   private String            identity;
 
@@ -59,7 +61,7 @@
   }
 
   static public String[] parse(String pstring) {
-    StringTokenizer parser = new StringTokenizer(pstring, AccessControlList.DELIMITER);
+    StringTokenizer parser = new StringTokenizer(pstring, IDENTITY_DELIMITER);
     String identityString = parser.nextToken();
     String permissionString = parser.nextToken();
 
@@ -93,7 +95,7 @@
 
   static protected String asString(String identity, String permission) {
     if (identity != null || permission != null) // SystemIdentity.ANY, PermissionType.ALL
-      return (identity != null ? identity : SystemIdentity.ANY) + AccessControlList.DELIMITER
+      return (identity != null ? identity : SystemIdentity.ANY) + IDENTITY_DELIMITER
           + (permission != null ? permission : PermissionType.READ);
     else
       return "";
